7.0 Doc Work (#3010)
GregTechCEu/GregTechCEu-1.19: docs/content/Modpacks/Recipes/CraftingComponents.md

Reviewer: for consistency, name this file `Crafting-Components.md`

---
title: "Crafting Components"
---

# Crafting Components

> Crafting Components are a way to organize and simplify the various similar recipes that GregTech generates. For example, writing out the recipes for all tiers of an alloy smelter can be done iteratively rather than hand written.

Reviewer: ```suggestion
Crafting Components are a way to organize and simplify the various similar recipes that GregTech generates. For example: writing out the recipes for all tiers of an alloy smelter can be done iteratively rather than one by one.
```

> Crafting Components are a map, pairing a `Voltage Tier` (the tier number) to a value. The valid value can either be a `MaterialEntry`, `ItemStack`, or a `TagPrefix<Item>`.

Reviewer: Voltage Tier has no reason to be an inline code block.
```suggestion
Crafting Components are a map pairing a Voltage tier (the tier number) to a value. The value can be a `MaterialEntry`, `ItemStack`, or `TagPrefix<Item>`.
```

## Changing a single entry

With KubeJS it is possible to modify the predefined components of existing GTCEu Modern machine crafting recipes.
You can replace singular entries, or do bulk modification of components.
-1 is defined as a fallback value if no other entries exist.

> ```js title="modification.js"

Reviewer: ```suggestion
```js title="startup/modification.js"
```

> const Map = Java.loadClass("java.util.Map")

Reviewer: does it still require a java Map? I thought I fixed that.

> GTCEuServerEvents.craftingComponents(event => {

Reviewer: it's a startup event.

    event.modifyItem(GTCraftingComponent.CIRCUIT, GTValues.MV, Item.of('minecraft:dirt')) // (1)
    event.modifyItem(GTCraftingComponent.PUMP, Map.of(
        GTValues.FALLBACK, Item.of('gtceu:lv_robot_arm'),
        GTValues.LV, Item.of('gtceu:lv_robot_arm'),
        GTValues.MV, Item.of('gtceu:mv_robot_arm'),
        GTValues.HV, Item.of('gtceu:hv_robot_arm'),
        GTValues.EV, Item.of('gtceu:ev_robot_arm'),
        GTValues.IV, Item.of('gtceu:iv_robot_arm'),
        GTValues.LuV, Item.of('gtceu:luv_robot_arm'),
        GTValues.ZPM, Item.of('gtceu:zpm_robot_arm'),
        GTValues.UV, Item.of('gtceu:uv_robot_arm'),
    )) // (2)
    event.modifyTag(CraftingComponent.CASING, GTValues.EV, 'minecraft:logs') // (3)

>     event.modifyMaterialEntry(CraftingComponent.PLATE, GTValues.UEV, new UnificationEntry('plate', 'gtceu:infinity')) // (4)

Reviewer: you're mixing both old variants and neither are correct.
you can access the registered crafting components with `GTCraftingComponents`, not `GTCraftingComponent` or `CraftingComponent`.

## Creating new components

It's also possible to create new crafting components with KubeJS.
The crafting component is constructed with a id and a fallback value. You can add entries by chaining `.add(tier, value)` methods after your construction.

```js title="creation.js"
const Map = Java.loadClass("java.util.Map")

let ITEM_CRAFTING_COMPONENT = null
let TAG_CRAFTING_COMPONENT = null
let UNIFICATION_CRAFTING_COMPONENT = null

GTCEuServerEvents.craftingComponents(event => {
    ITEM_CRAFTING_COMPONENT = event.createItem("item_component", Item.of('minecraft:cyan_stained_glass'), Map.of(
        GTValues.LV: Item.of('minecraft:cyan_stained_glass'),
        GTValues.MV: Item.of('minecraft:cyan_stained_glass'),
        GTValues.HV: Item.of('minecraft:cyan_stained_glass'),
        GTValues.EV: Item.of('minecraft:lime_stained_glass'),
        GTValues.IV: Item.of('minecraft:lime_stained_glass'),
        GTValues.LuV: Item.of('minecraft:lime_stained_glass'),
        GTValues.ZPM: Item.of('minecraft:magenta_stained_glass'),
        GTValues.UV: Item.of('minecraft:magenta_stained_glass'),

>     )) // (1)

Reviewer: the `key: value` syntax does not work with Map.of(), as that's the java method and not the KJS wrapper.

    TAG_CRAFTING_COMPONENT = event.createTag("tag_component", 'forge:barrels/wooden', Map.of(
        GTValues.LV: 'forge:chests/wooden',
        GTValues.MV: 'forge:chests/trapped',
        GTValues.HV: 'forge:chests/ender',
        GTValues.EV: 'forge:cobblestone',
        GTValues.IV: 'forge:cobblestone/normal',
        GTValues.LuV: 'forge:cobblestone/infested',
        GTValues.ZPM: 'forge:cobblestone/mossy',
        GTValues.UV: 'forge:cobblestone/deepslate',

>     )) // (2)

Reviewer: same here

    UNIFICATION_CRAFTING_COMPONENT = event.createMaterialEntry("material_entry_component", new UnificationEntry('plate', 'gtceu:infinity'), Map.of(
        GTValues.LV: new MaterialEntry('block', 'gtceu:infinity'),
        GTValues.MV: new MaterialEntry('ingot', 'gtceu:infinity'),
        GTValues.HV: new MaterialEntry('dust', 'gtceu:infinity'),
        GTValues.EV: new MaterialEntry('round', 'gtceu:infinity'),
        GTValues.IV: new MaterialEntry('foil', 'gtceu:infinity'),
        GTValues.LuV: new MaterialEntry('longRod', 'gtceu:infinity'),
        GTValues.ZPM: new MaterialEntry('rod', 'gtceu:infinity'),

>         GTValues.UV: new MaterialEntry('bolt', 'gtceu:infinity'),

Reviewer: same here
